### DISCUSSION TITLE
Difference between yul : true and viaIR : true

### DISCUSSION THREAD

**Discussion by clauBv23** [NomicFoundation/hardhat#4099] [Troubleshooting] [answered]
I have always used `viaIR:true` to enable the yul optimizer, which has worked so far. 

But it is not clearly explained in the [solidity configuration documentation](https://hardhat.org/hardhat-runner/docs/config#solidity-configuration).
Instead, it says `settings` could be configured via input JSON as defined in the solidity documentation, 
pointing up that: 
 ```  // The new Yul optimizer. Mostly operates on the code of ABIEncoderV2
        // and inline assembly.
        // It is activated together with the global optimizer setting
        // and can be deactivated here.
        // Before Solidity 0.6.0 it had to be activated through this switch.
        "yul": false,
```

So, how should be enabled the yul optimizer:
with `viaIR`
``` 
settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
      viaIR: true, 
    },
```
or with the input JSON format 

```
settings: {
      optimizer: {
        enabled: true,
        runs: 200,
        details: {
          yul: true,
        },
      },
    },
```
I tried the second one( version `0.8.20`) and didn't work, does it only work for versions before `0.6.0`?

**Comment by fvictorio** [comment]
Yeah, I think `viaIR: true` is the correct way to do it.

My understanding of that comment is that you can use `yul: false` on top of an enabled optimizer to disable a specific optimization... but I'm not really sure.

@cameel sorry for the tag, but could you clarify what this means?

**Comment by cameel** [comment, marked as answer]
Sure, not problem.

`viaIR` enables the new code generation pipeline. So this makes the compiler first generate Yul and only then EVM bytecode (rather than going straight from Solidity to EVM assembly).

`optimizer.details.yul` enables Yul optimizer, which affects Yul generated by the new pipeline but also Yul generated in any other circumstances. For example you can request `irOptimized` output even when using the legacy code generation pipeline - that output will be passed through the Yul optimizer when it is enabled. Also, some language features, even in the legacy pipeline, are generated from Yul code - e.g. anything that uses abicoder v2.

So, basically, if you want fully optimized Yul output, you need both. If you only use `viaIR: true`, the compiler will generate Yul code and generate EVM bytecode from that without optimization. That EVM bytecode may still be optimized by the legacy EVMasm optimizer if you set `optimizer.enabled: true`, but that's less powerful.

**Reply by clauBv23** [reply]
Perfect! Thanks for the detailed explanation 🙌 , just a quick question to clarify.
```
        // The new Yul optimizer. Mostly operates on the code of ABIEncoderV2
        // and inline assembly.
        // It is activated together with the global optimizer setting
        // and can be deactivated here.
        // Before Solidity 0.6.0 it had to be activated through this switch.
        "yul": false, 
```
When sais:

> // It is activated together with the global optimizer setting
        // and can be deactivated here.

Does that mean that enabling the optimizer will also enable yul optimizer, or should `optimizer.details.yul` be explicitly enabled?

I mean, going with:
```
settings: {
      optimizer: {
        enabled: true,
        runs: 200,
      },
      viaIR: true, 
    },
```
will be enough Or should add
``` 
        details: {
           yul: true,
        },
```

**Reply by cameel** [reply]
> Does that mean that enabling the optimizer will also enable yul optimizer

Yes. The main optimizer setting enables all of the optimizer components (i.e. EVM asm optimizer and Yul optimizer). You can then use more specific settings to selectively disable these components or their steps. When the main optimizer setting is disabled, you can use them to selectively enable components/steps.

Overall, this is the minimal set of settings that enables full optimization and the new pipeline:
```javascript
settings: {
    optimizer: {enabled: true},
    viaIR: true, 
},
```

**Reply by clauBv23** [reply]
Perfect!!!/ Thanks for your time
